cpmpy/solvers/cpo.py

```diff
@@ -0,0 +1,470 @@
+                    print(argvals(display))
+                else:
+                    display()  # callback
+        return solution_count
+
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        docp = self.get_docp()
+        if is_num(cpm_var): # shortcut, eases posting constraints
+            return cpm_var
+
+        # special case, negative-bool-view
+        # work directly on var inside the view
+        if isinstance(cpm_var, NegBoolView):
+            return (self.solver_var(cpm_var._bv) == 0)
+
+        # create if it does not exist
+        if cpm_var not in self._varmap:
+            if isinstance(cpm_var, _BoolVarImpl):
+                # note that a binary var is an integer var with domain (0,1), you cannot do boolean operations on it.
+                # we should add == 1 to turn it into a boolean expression
+                revar = docp.expression.binary_var(str(cpm_var)) == 1
+            elif isinstance(cpm_var, _IntVarImpl):
+                revar = docp.expression.integer_var(min=cpm_var.lb, max=cpm_var.ub, name=str(cpm_var))
+            else:
+                raise NotImplementedError("Not a known var {}".format(cpm_var))
+            self._varmap[cpm_var] = revar
+
+        # return from cache
+        return self._varmap[cpm_var]
+
+    def objective(self, expr, minimize=True):
+        """
+            Post the given expression to the solver as objective to minimize/maximize
+
+            'objective()' can be called multiple times, only the last one is stored
+
+            (technical side note: any constraints created during conversion of the objective
+
+            are permanently posted to the solver)
+        """
+        dom = self.get_docp().modeler
+        if self.has_objective():
+            self.cpo_model.remove(self.cpo_model.get_objective_expression())
+        expr = self._cpo_expr(expr)
+        if minimize:
+            self.cpo_model.add(dom.minimize(expr))
+        else:
+            self.cpo_model.add(dom.maximize(expr))
+
+    def has_objective(self):
+        return self.cpo_model.get_objective() is not None
+
+    # `__add__()` first calls `transform()`
+    def transform(self, cpm_expr):
+        """
+            Transform arbitrary CPMpy expressions to constraints the solver supports
+
+            Implemented through chaining multiple solver-independent **transformation functions** from
+            the `cpmpy/transformations/` directory.
+
+            See the 'Adding a new solver' docs on readthedocs for more information.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: list of Expression
+        """
+        # apply transformations
+        cpm_cons = toplevel_list(cpm_expr)
+        # count is only supported with a constant to be counted, so we decompose
```

Comment:
what about the new no_partial_functions? or can it handle that?

Reply:
ah yes definitely should be added!
I guess it would be good to add some testcases to our test_solvers to check for this. Test will also make clear whether we need to safen toplevel for div/mod